### ISSUE TITLE
Tuxemon crashes when some settings aren't present in tuxemon.cfg

### ISSUE THREAD

**Issue by MirceaKitsune** [Tuxemon/Tuxemon#382]
Tuxemon refuses to start when certain settings are missing from the user's tuxemon.cfg file. This is especially an issue for those who run the development version, as they have to regenerate their config or manually add some new options.

As an example, I removed net_controller_enabled from my "~/.tuxemon/tuxemon.cfg" file. Instead of ignoring this and using a default value, Tuxemon crashes with the following output:

```
mircea@linux-qz0r:~/Games/Small/Python/Tuxemon/Tuxemon_GIT/tuxemon> python ./tuxemon.py
Traceback (most recent call last):
  File "./tuxemon.py", line 60, in <module>
    from core.main import main
  File "/home/mircea/Games/Small/Python/Tuxemon/Tuxemon_GIT/tuxemon/core/main.py", line 33, in <module>
    from . import prepare
  File "/home/mircea/Games/Small/Python/Tuxemon/Tuxemon_GIT/tuxemon/core/prepare.py", line 77, in <module>
    CONFIG = config.Config(CONFIG_FILE_PATH)
  File "/home/mircea/Games/Small/Python/Tuxemon/Tuxemon_GIT/tuxemon/core/components/config.py", line 71, in __init__
    self.net_controller_enabled = self.config.get("game", "net_controller_enabled")
  File "/usr/lib64/python2.7/ConfigParser.py", line 618, in get
    raise NoOptionError(option, section)
ConfigParser.NoOptionError: No option 'net_controller_enabled' in section: 'game'
mircea@linux-qz0r:~/Games/Small/Python/Tuxemon/Tuxemon_GIT/tuxemon> 
```

The solution seems pretty obvious: All options in tuxemon/core/components/config.py should be defined using the try / except Python functions. This is done for newer settings, only older ones expect the config to be present at all times. An example of how this is done right:

```
try:
    self.locale = self.config.get("game", "locale")
except configparser.NoOptionError:
    self.locale = "en_US"
```

@ShadowApex @bitcraft Do you agree that doing this for all entries is the correct solution? If so I'll make a pull request and fix all of the settings listed there.

**Comment by bitcraft** [comment]
The configparser object can accept a dictionary which defines the defaults.  IMO, the correct solution would be to define a dictionary that are the defaults, then pass that to the constructor, since that is a feature built into the class.

 https://docs.python.org/2/library/configparser.html

I haven't tried this, but it may/should be possible to read the configuration from the tuxemon game folder and then either pass that to configparser object that reads the user configuration, or to build a dictionary from the default configuration, then pass that to the configparser that reads the user config.  That way, it would be possible to define the defaults by the configuration file that is bundled with the game.

https://github.com/Tuxemon/Tuxemon/blob/master/tuxemon/core/components/config.py#L47

I'd rather not clutter up the module with a bunch of try/excepts.

**Comment by MirceaKitsune** [comment]
Hmmm. To be fair I'm a little surprised that config.get wasn't itself designed with a 3rd parameter that can act as a default... instead Python normally prefers to abort on missing options altogether. I'm also seeing a config.has_option function, which could alternatively be used to check for a setting before using config.get... of course this wouldn't fix the clutter.

As far as a dictionary goes, I'm seeing there that you need to use config.defaults to define one. Indeed it would be preferable if the tuxemon.cfg included in Git could be used for this purpose, rather than a hardcoded settings list which is recursive. Any example on how exactly I'd have to write and format that function to get it working?

**Comment by bitcraft** [comment]
I just went through the docs.  I'm honestly amazed that they did not implement the default as a dictionary with sections as dictionaries.  Nested...  Seems really short sighted.  

If I were going to do this, I would do one of the following...not sure which.

1. Make new function which will read one ConfigParser obj and read defaults from another
2. Subclass ConfigParser and override ConfigParser.get to get defaults from another
3. Remove the sample config file, rewrite it using ConfigParser.set, do one of the above
4. Use ConfigObj
5. Use YAML, implement defaults using dict.update()

For options #1 and #2, you will need to read the all the options from the defaults, but set values from the config first, then use defaults if they are not found.

ConfigParser.get can accept a default value, but it is not really intuitive, IMO.  you need to give it an object (like dict) to get the missing key/value pair from.

Option #1 would probably be the easiest.

```
# implemented as a closure
def default_getter(config, defaults):
   def func(section, name):
      try:
         return config.get(section, name)
      except NoSectionError:
         return defaults.get(section, name)
   return func

config = ConfigParser(...)
defaults = ConfigParser(...)

getter = default_getter(config, defaults)

self.starting_map = getter("game", "starting_map")
```

If it were me, I'd also take it further by creating validators and removing all hard references to options from the python source, and iterate over the defaults' sections and values to build the CONFIG object.

Just my thoughts.

**Comment by MirceaKitsune** [comment]
I provided a fix in #384 Settings missing from the user's tuxemon.cfg will now be read from the default tuxemon.cfg. I also reordered all settings accordingly, removing hardcoded defaults as well.

Settings themselves are still defined manually; I decided not to bother with the Python config dictionary and using a loop... too complicated and too big of a change for now. This model is however good, as it guarantees the code will ignore useless settings and not fill the memory up with them.
